hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/ClientInvocationServiceImpl.java

```diff
@@ -294,6 +294,36 @@ void checkInvocationAllowed() throws IOException {
         connectionManager.checkInvocationAllowed();
     }
 
+    void checkUrgentInvocationAllowed(ClientInvocation invocation) {
+        if (!client.shouldCheckUrgentInvocations()) {
+            // If there were no Compact schemas to begin with, we don't need
+            // to perform the checks below. If the client didn't send a Compact
+            // schema up until this point, the retries or listener registrations
+            // could not send a schema, because if they were, we wouldn't hit
+            // this line.
+            return;
```

Comment:
what about concurrency? What happens when you return here, at the same time a new compact schema is populated into the list of schemas?

Reply:
I think it won't be a problem. I have thought it as well. toData is called first before sending the invocation. If toData did not add a schema to the SchemaService then it means data is not compact serializable or toDataWithSchema is used. 

Another schema can be added right after this check and it won't be a problem because this invocation is not related with that schema.

Reply:
Let's think of this:
- For retries, we are safe, because for an invocation to be retried, it must be invoked first, and if it is invoked and it has some compact serialized data, the schema associated with that should be replicated to the cluster and put into client's local registry.

- For urgent invocations like listener registrations, which can include compact data, the same applies. For them to be stored in the listener service, they must be invoked first, and for them to be invoked, the schema associated with that invocation must be replicated.

The idea here is that, when I do this check, if there were a Compact schema associated with this invocation, then there must be some schema in client's local registry. Of course, some new schemas might be added concurrently, but they are for sure not related to this invocation, so we are safe.